text/0195-associated-items.md

```diff
@@ -1190,6 +1197,73 @@ trait Mappable
 While the above demonstrates the versatility of associated types and `where`
 clauses, it is probably too much of a hack to be viable for use in `libstd`.
 
+### Associated consts in generic code
+
+There are some restrictions on uses of associated consts in generic code. These
+might be loosened or removed in the future (see the related sub-sections in
+"Unresolved questions" below).
+
+ 1. Values of constant expressions in match patterns cannot depend on a type
+    parameter (by extension, neither can the types of such expressions). This
+    restriction is necessary for exhaustiveness and reachability to be checked
+    in generic code.
+
+    Note that the dependence of a value on a type parameter may be indirect:
+
+    ```rust
+    enum MyEnum {
+        Var1,
+        Var2,
+    }
+    trait HasVar {
+        const VAR: MyEnum;
+    }
+    fn do_something<T: HasVar>(x: MyEnum) {
+        const y: MyEnum = <T>::VAR;
```

Comment:
Oh, plus, if you want to have associated consts depend on type parameters in traits, you already have to do most of the implementation work, and AFAIK, there's no additional reason not to allow it in functions at that point. What I mean is that we presumably expect to allow impls like this:

```rust
struct WrapInt<T: Int>(T);

impl<T: Int> Int for WrapInt<T> {
    const ZERO: WrapInt<T> = WrapInt(<T as Int>::ZERO);
}
```

(Well, maybe that *exact* example is not so great in that `Int` is designed for primitives, but you get the idea.)